### PR TITLE
util: use generics to implement max/min

### LINKED ISSUE #34247: use Generics to unify Max/Min

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: close #34247

Problem Summary:

### What is changed and how it works?

- remove ```github.com/cznic/mathutil``` and ```modernc.org/mathutil```
- rename ```util/math``` to ```util/mathutil```
- upgrade the linter to adapt generics

### Check List

Tests 

- [x] Unit test
- [x] Integration test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/utils/math_test.go
Comment (siddontang) [inline br/pkg/utils/math_test.go:13]: if you implement this in util/math, please move all related tests to util/math_test.go, not in br folder now.

File: util/math/math.go
Comment (disksing) [inline util/math/math.go:84]: why not merge the 2 methods? I think the multi version can accept 2 arguments too.

File: util/math/math.go
Comment (disksing) [inline util/math/math.go:60]: consider use `golang.org/x/exp/constraints.Ordered`?

Comment (hawkingrei) [reply util/math/math.go]: Done

File: util/math/math.go
Comment (hawkingrei) [reply util/math/math.go]: Done

File: br/pkg/utils/math_test.go
Comment (hawkingrei) [reply br/pkg/utils/math_test.go]: Done

File: util/math/math.go
Comment (disksing) [inline util/math/math.go:62]: It can work with non-integers too.

File: br/pkg/utils/math.go
Comment (disksing) [inline br/pkg/utils/math.go:17]: Consider making `Clamp` a generic function too?

File: util/math/math.go
Comment (hawkingrei) [reply util/math/math.go]: Done

File: br/pkg/utils/math.go
Comment (hawkingrei) [reply br/pkg/utils/math.go:17]: @disksing Yes, I will. But it needs to unite the other functions into a generic function. it will be a big PR. so I will create a new pr to make it.

**Comment by hawkingrei** [comment]
linters have been updated for adapting golang 1.18.

**Comment by hawkingrei** [comment]
@siddontang @disksing @tisonkun PTAL

File: go.mod
Comment (tisonkun) [inline go.mod:98]: Merge to above block?

File: planner/core/common_plans.go
Comment (tisonkun) [inline planner/core/common_plans.go:774]: Unrelated change?

File: planner/core/logical_plan_builder.go
Comment (tisonkun) [inline planner/core/logical_plan_builder.go:4654]: Unrelated change?

File: statistics/cmsketch.go
Comment (tisonkun) [inline statistics/cmsketch.go:374]: Unrelated change?

File: go.mod
Comment (hawkingrei) [reply go.mod:98]: It is from the ```util/math```. I restart to format it and put it into the front part together.

File: planner/core/common_plans.go
Comment (hawkingrei) [reply planner/core/common_plans.go:774]: It is related with the linter upgrading.

Comment (hawkingrei) [reply planner/core/common_plans.go:774]: It is related with the linter upgrading.

File: planner/core/logical_plan_builder.go
Comment (hawkingrei) [reply planner/core/logical_plan_builder.go:4654]: linter upgrading for Golang 1.18 generics.

**Review by tisonkun — APPROVED** [review]

File: expression/builtin_control.go
Comment (mjonss) [inline expression/builtin_control.go:23]: Maybe we should be consistent and use either a unique package name, so we don't need an alias, or use a unique alias so it is the same everywhere?
So far I have seen 'math', 'tidbmath' and 'mathutil'. Where 'math' is too easy to be taken as the standard 'math' package.
I propose to have the name mathutil, but are open to other names as well, just as long as it is not colliding with others and can be used consistently.

File: util/math/math_test.go
Comment (mjonss) [inline util/math/math_test.go:71]: Is this really testing Int64? In that case what is the TestMinInt testing ;)
Maybe have tests for all different int types and why not also for other constraints.Ordered types, like float and string types?

**Review by mjonss — COMMENTED** [review]
What are the benefits apart from using a new cool feature?

Have you done any benchmarking if the generics are faster in this case?

File: util/math/math_test.go
Comment (hawkingrei) [reply util/math/math_test.go:71]: This is indeed a problem caused by migration, let me refactor it.

Comment (hawkingrei) [reply util/math/math_test.go]: Put all tests of Max/Min into a test case.

**Comment by hawkingrei** [comment]
> What are the benefits apart from using a new cool feature?
> 
> Have you done any benchmarking if the generics are faster in this case?

1、The main purpose is to strengthen the base library and remove duplicate code. so I disagree that generics can be used everywhere in the TiDB repo.  Golang's generics are not very mature, so we'd better just use them simply to gain experience.

2、I have add the benchmark. Here is the benchmark code.

```
package main

import (
	"math/rand"
	"testing"
	"time"
)

func BenchmarkMaxGenerics(b *testing.B) {
	rand.Seed(time.Now().Unix())
	for n := 0; n < b.N; n++ {
		Max(1, rand.Perm(10)...) // run fib(30) b.N times
	}
}

func BenchmarkMax(b *testing.B) {
	rand.Seed(time.Now().Unix())
	for n := 0; n < b.N; n++ {
		MinInt(1, rand.Perm(10)...) // run fib(30) b.N times
	}
}

```

Here is the result.

```
> $ go test -bench=. -benchtime=5s -count=3 .                                                                                                                     
goos: darwin
goarch: amd64
pkg: gomain
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkMaxGenerics-16         23212767               256.9 ns/op
BenchmarkMaxGenerics-16         23165785               258.9 ns/op
BenchmarkMaxGenerics-16         22609570               281.8 ns/op
BenchmarkMax-16                 20863150               294.3 ns/op
BenchmarkMax-16                 20165695               302.9 ns/op
BenchmarkMax-16                 19426915               309.4 ns/op
PASS
ok      gomain  38.443s
```

It seems that the generic one has better performance.

**Comment by hawkingrei** [comment]
```
> $ go test -bench=. -benchtime=10s .                                                                                                                             
goos: darwin
goarch: amd64
pkg: gomain
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkMaxGenerics-16         34395898               346.1 ns/op
BenchmarkMax-16                 34346350               342.2 ns/op
PASS
ok      gomain  24.556s
```

update benchmark result. Their performance is about the same.

File: br/pkg/lightning/mydump/router.go
Comment (mjonss) [inline br/pkg/lightning/mydump/router.go:124]:
```suggestion
	// Route apply rule to path. Return nil if path doesn't match route rule;
```

File: br/pkg/lightning/mydump/router_test.go
Comment (mjonss) [inline br/pkg/lightning/mydump/router_test.go:163]:
```suggestion
	// add another rule that match same pattern with the third rule, the result should be no different
```

File: br/pkg/lightning/restore/check_info.go
Comment (mjonss) [inline br/pkg/lightning/restore/check_info.go:54]: I think you can remove the dependency of modernc.org/mathutil and replace it with github.com/pingcap/tidb/util/mathutil instead since only some Max/Min functions was used from it anyway.

I could only find it to be used in these files:
br/pkg/lightning/checkpoints/checkpoints.go
br/pkg/lightning/restore/check_info.go
br/pkg/lightning/restore/restore.go

File: br/pkg/lightning/restore/restore.go
Comment (mjonss) [inline br/pkg/lightning/restore/restore.go:60]: Since you removed mathutil below, no need to create an alias here

File: ddl/index_modify_test.go
Comment (mjonss) [inline ddl/index_modify_test.go:212]: Please revert this, math.MaxInt64 is a constant here I believe.

File: ddl/sequence.go
Comment (mjonss) [inline ddl/sequence.go:29]: I would prefer to remove the dependency of github.com/cznic/mathutil and replace the Max/Min functions from there with github.com/pingcap/tidb/util/mathutil and then there will be no need for the alias.

File: ddl/sequence_test.go
Comment (mjonss) [inline ddl/sequence_test.go:60]: Please revert, the math.MaxInt64 is correct here.

File: ddl/sequence_test.go
Comment (mjonss) [inline ddl/sequence_test.go:1034]: Please revert, the math.MaxInt64 is correct here.

File: executor/point_get_test.go
Comment (mjonss) [inline executor/point_get_test.go:807]: Please revert this, Please revert, the math.MaxUint64 is correct here.

File: executor/split.go
Comment (mjonss) [inline executor/split.go:233]: Please revert, the math.{Max,Min}Int64 is correct here.

File: expression/builtin_arithmetic.go
Comment (mjonss) [inline expression/builtin_arithmetic.go:27]: Please remove the alias and github.com/cznic/mathutil dependency and use the generics from github.com/pingcap/tidb/util/mathutil instead.

File: expression/builtin_control.go
Comment (mjonss) [inline expression/builtin_control.go:23]:
```suggestion
	"github.com/pingcap/tidb/util/mathutil"
```

File: expression/builtin_convert_charset.go
Comment (mjonss) [inline expression/builtin_convert_charset.go:293]: Please revert this, since I think it means the expression/builtin_math dependency.

File: expression/builtin_math.go
Comment (mjonss) [inline expression/builtin_math.go:34]: Please remove the dependency of github.com/cznic/mathutil, and replace the constants with constants from standard library "math" and Max/Min with the new mathutil package and remove the alias.

File: expression/builtin_math.go
Comment (mjonss) [inline expression/builtin_math.go:657]: Please remove, the comment is still correct.

File: expression/builtin_math_test.go
Comment (mjonss) [inline expression/builtin_math_test.go:30]: No need for an alias here, please use mathutil instead.

File: expression/builtin_math_test.go
Comment (mjonss) [inline expression/builtin_math_test.go:832]: Please revert, I think the comment is still correct.

File: expression/builtin_math_vec.go
Comment (mjonss) [inline expression/builtin_math_vec.go:25]: Please replace the usage of github.com/cznic/mathutil with github.com/pingcap/tidb/util/mathutil and remove the utilMath alias.

File: expression/builtin_op.go
Comment (mjonss) [inline expression/builtin_op.go:826]: I think you can revert these changes, since the comments are most likely still correct.

File: expression/expression.go
Comment (mjonss) [inline expression/expression.go:962]: I think you can revert these comment changes as well

File: expression/generator/time_vec.go
Comment (mjonss) [inline expression/generator/time_vec.go:715]: Is this correct?

File: parser/ast/functions.go
Comment (mjonss) [inline parser/ast/functions.go:84]: Please revert this.

File: planner/core/common_plans.go
Comment (mjonss) [inline planner/core/common_plans.go:1658]: Please revert these comment changes.

File: sessionctx/variable/session.go
Comment (mjonss) [inline sessionctx/variable/session.go:49]: Please remove the alias and update the file to use the mathutil namespace instead.

File: store/mockstore/unistore/cophandler/cop_handler_test.go
Comment (mjonss) [inline store/mockstore/unistore/cophandler/cop_handler_test.go:320]: Please revert this comment change.

File: table/tables/partition.go
Comment (mjonss) [inline table/tables/partition.go:187]: Please revert this comment change.

File: table/tables/partition.go
Comment (mjonss) [inline table/tables/partition.go:423]: Please revert this comment change.

File: table/temptable/interceptor_test.go
Comment (mjonss) [inline table/temptable/interceptor_test.go:80]: Please revert this comment change.

File: types/convert.go
Comment (mjonss) [inline types/convert.go:175]: Please revert this comment change.

File: types/field_type.go
Comment (mjonss) [inline types/field_type.go:27]: Please remove the alias

File: types/mydecimal.go
Comment (mjonss) [inline types/mydecimal.go:1124]: Please revert this comment change.

File: util/hack/hack.go
Comment (mjonss) [inline util/hack/hack.go:51]: Please revert this comment change.

File: util/mathutil/math.go
Comment (mjonss) [inline util/mathutil/math.go:37]: Please revert this comment change.

**Review by mjonss — APPROVED** [review]
Some automatic replace that updated comments and a few places where alias is no longer needed.

Otherwise LGTM. Please fix before merge :)

File: ddl/sequence.go
Comment (hawkingrei) [reply ddl/sequence.go]: removed ```modernc.org/mathutil``` and ```github.com/cznic/mathutil```.

File: expression/builtin_arithmetic.go
Comment (hawkingrei) [reply expression/builtin_arithmetic.go:27]: Done

File: expression/builtin_control.go
Comment (hawkingrei) [reply expression/builtin_control.go:23]: Done

File: expression/builtin_math.go
Comment (hawkingrei) [reply expression/builtin_math.go]: Done

File: expression/builtin_control.go
Comment (hawkingrei) [reply expression/builtin_control.go]: I have renamed ```util/math``` to ```util/mathutil```. and remove ```mathutil``` third library.

File: expression/generator/time_vec.go
Comment (hawkingrei) [reply expression/generator/time_vec.go]: reverted.

File: expression/builtin_math_test.go
Comment (hawkingrei) [reply expression/builtin_math_test.go]: Done

File: util/mathutil/math.go
Comment (bb7133) [inline util/mathutil/math.go:70]:
```suggestion
// Max returns the largest one from its arguments.
```

File: util/mathutil/math.go
Comment (bb7133) [inline util/mathutil/math.go:81]:
```suggestion
// Min returns the smallest one from its arguments.
```
Please make sure the comments of `Max` and `Min` follows the similar style.

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**
